flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/windows.scala

```diff
@@ -22,6 +22,126 @@ import org.apache.flink.table.expressions.{Expression, ExpressionParser}
+  *
+  * Over window is similar to the traditional OVER SQL.
+  */
+class OverWindow {
+
+  private[flink] var alias: Expression = _
+  private[flink] var partitionBy: Seq[Expression] = Seq[Expression]()
+  private[flink] var orderBy: Expression = _
+  private[flink] var preceding: Expression = _
+  private[flink] var following: Expression = null
+
+  /**
+    * Assigns an alias for this window that the following `select()` clause can refer to.
+    *
+    * @param alias alias for this over window
+    * @return this over window
+    */
+  def as(alias: String): OverWindow = as(ExpressionParser.parseExpression(alias))
+
+  /**
+    * Assigns an alias for this window that the following `select()` clause can refer to.
+    *
+    * @param alias alias for this over window
+    * @return this over window
+    */
+  def as(alias: Expression): OverWindow = {
+    this.alias = alias
+    this
+  }
+
+  /**
+    * Partitions the elements on some partition keys.
+    *
+    * @param partitionBy
+    * @return this over window
+    */
+  def partitionBy(partitionBy: String): OverWindow = {
+    this.partitionBy(ExpressionParser.parseExpression(partitionBy))
+  }
+
+  /**
+    * Partitions the elements on some partition keys.
+    *
+    * @param partitionBy
+    * @return this over window
+    */
+  def partitionBy(partitionBy: Expression*): OverWindow = {
+    this.partitionBy = partitionBy
+    this
+  }
+
+
+  /**
+    * Specifies the time mode.
+    *
+    * @param orderBy For streaming tables call [[orderBy 'rowtime or orderBy 'proctime]]
+    *                to specify time mode.
+    * @return this over window
+    */
+  def orderBy(orderBy: String): OverWindow = {
+    this.orderBy(ExpressionParser.parseExpression(orderBy))
+  }
+
+  /**
+    * Specifies the time mode.
+    *
+    * @param orderBy For streaming tables call [[orderBy 'rowtime or orderBy 'proctime]]
+    *                to specify time mode.
+    * @return this over window
+    */
+  def orderBy(orderBy: Expression): OverWindow = {
+    this.orderBy = orderBy
+    this
+  }
+
+  /**
+    * Set the preceding offset (based on time or row-count intervals) for over window
+    *
+    * @param preceding forward offset that relative to the current row
+    * @return this over window
+    */
+  def preceding(preceding: String): OverWindow = {
+    this.preceding(ExpressionParser.parseExpression(preceding))
+  }
+
+  /**
+    * Set the preceding offset (based on time or row-count intervals) for over window
+    *
+    * @param preceding forward offset that relative to the current row
+    * @return this over window
+    */
+  def preceding(preceding: Expression): OverWindow = {
+    this.preceding = preceding
+    this
+  }
+
+  /**
+    * Set the following offset (based on time or row-count intervals) for over window
+    *
+    * @param following subsequent offset that relative to the current row
+    * @return this over window
+    */
+  def following(following: String): OverWindow = {
```

Comment:
Should we make `following` optional and default to CURRENT_ROW / CURRENT_RANGE depending on the type of `preceding`? I think that would be a nice shortcut and also be aligned with SQL.

What do you think @sunjincheng121 ?